calico_node/startup/startup.go

```diff
@@ -0,0 +1,600 @@
+			terminate()
+		}
+	} else {
+		if ipv4Env != "" {
+			node.Spec.BGP.IPv4Address = parseIPEnvironment("IP", ipv4Env, 4)
+		}
+		validateIP(node.Spec.BGP.IPv4Address)
+	}
+
+	ipv6Env := os.Getenv("IP6")
+	if ipv6Env == "autodetect" {
+		adm := os.Getenv("IP6_AUTODETECT_METHOD")
+		node.Spec.BGP.IPv6Address = autoDetectCIDR(adm, 6)
+	} else {
+		if ipv6Env != "" {
+			node.Spec.BGP.IPv6Address = parseIPEnvironment("IP6", ipv6Env, 6)
+		}
+		validateIP(node.Spec.BGP.IPv6Address)
+	}
+
+}
+
+// fetchAndValidateIPAndNetwork fetches and validates the IP configuration from
+// either the environment variables or from the values already configured in the
+// node.
+func parseIPEnvironment(envName, envValue string, version int) *net.IP {
+	ip := &net.IP{}
+	err := ip.UnmarshalText([]byte(envValue))
+	if err != nil || ip.Version() != version {
+		fatal("Environment does not contain a valid IPv%d address: %s=%s", version, envName, envValue)
+		terminate()
+	}
+	message("Using IPv%d address from environment: %s=%s", ip.Version(), envName, envValue)
+
+	return ip
+}
+
+// validateIP checks that the IP address is actually on one of the host
+// interfaces and warns if not.
+func validateIP(ip *net.IP) {
+	// No validation required if no IP address is specified.
+	if ip == nil {
+		return
+	}
+
+	// Get a complete list of interfaces with their addresses and check if
+	// the IP address can be found.
+	ifaces, err := autodetection.GetInterfaces(nil, nil, ip.Version())
+	if err != nil {
+		fatal("Unable to query host interfaces: %s", err)
+		terminate()
+	}
+	if len(ifaces) == 0 {
+		message("No interfaces found for validating IP configuration")
+	}
+
+	for _, i := range ifaces {
+		for _, c := range i.Cidrs {
+			if ip.Equal(c.IP) {
+				message("IPv%d address %s discovered on interface %s", ip.Version(), ip.String(), i.Name)
+				return
+			}
+		}
+	}
+	warning("Unable to confirm IPv%d address %s is assigned to this host", ip.Version(), ip)
+}
+
+// autoDetectCIDR auto-detects the IP and Network using the requested
+// detection method.
+func autoDetectCIDR(detectionMethod string, version int) *net.IP {
+	incl := []string{}
+	excl := []string{"^docker.*", "^cbr.*", "dummy.*",
+		"virbr.*", "lxcbr.*", "veth.*", "lo",
+		"cali.*", "tunl.*", "flannel.*"}
+
+	// At the moment, we don't support anything other than the default
+	// (blank) auto-detection method.
+	if detectionMethod != "" {
+		fatal("IP detection method is not supported: %s", detectionMethod)
+		terminate()
+	}
+
+	iface, cidr, err := autodetection.FilteredEnumeration(incl, excl, version)
+	if err != nil {
+		message("Unable to auto-detect any valid IPv%d addresses: %s", version, err)
+		return nil
+	}
+
+	if cidr == nil {
+		message("Unable to auto-detect an IPv%d address", version)
+		return nil
+	}
+
+	message("Using autodetected IPv%d address on interface %s: %s", version, iface.Name, cidr.String())
+
+	return &net.IP{cidr.IP}
```

Comment:
I realize this is just returning the IP address at the moment -- but once the libcalico-go changes are in this will be updated to just return the CIDR as per the interface updates to the node resource.